### ISSUE TITLE
Minor issue in StringUtil::encodeEmail

### ISSUE THREAD

**Issue by fritzmg** [contao/core#7984]
The last part of this regular expression
```php
preg_match_all('/\w([-._\w]*\w)?@\w([-._\w]*\w)?\.\w{2,6}/', $strString, $arrEmails);
```
i.e. `.\w{2,6}` seems to suggest, that you wanted to only match email addresses whose TLD is between 2 and 6 characters long. This does not actually work:

![regex](https://cloud.githubusercontent.com/assets/4970961/9486873/d5f5f7ec-4bcc-11e5-8b17-71f29e79cddc.png)

_However_, it's good that it does not work, since there are TLDs now with more than 6 characters, but wanted to mention it nevertheless ;). (For example, if you use the insert tag `{{email::test@test.loremipsum}}` then it will still work, even though the last part of the regex supposedly only matches TLDs between 2 and 6 characters long.)

**Comment by leofeyer** [comment]
@contao/developers How would you solve this? Should we simply remove the length pattern `{2,6}`? If we did, would we have to validate the findings (e.g. using https://github.com/jeremykendall/php-domain-parser)?

**Comment by leofeyer** [comment]
BTW, we have the same problem here: https://github.com/contao/core/blob/master/system/modules/comments/classes/Comments.php#L457

**Comment by Toflar** [comment]
Big :+1: for using the mentioned library. Let's reuse whatever we can reuse :)

**Comment by aschempp** [comment]
:+1: for a library in general, but be aware that stuff like `.local` will no longer work, and I'm not sure if we should add a new dependency in a PATCH version…

**Comment by ausi** [comment]
IMO we should change the length pattern to `{2,63}` because 63 is the maximum length of a DNS label [according to Wikipedia](https://en.wikipedia.org/wiki/Domain_Name_System#Domain_name_syntax).

The regular expression also fails to match some other forms of email addresses:

```
foo+bar@gmail.com
foo@exämple.com
föö@example.com
```

**Comment by leofeyer** [comment]
As discussed on Mumble on September 10th, we should increase the pattern length to 63 as proposed by @ausi [above](https://github.com/contao/core/issues/7984#issuecomment-138589449). Maybe we can switch to a library in Contao 4.

**Comment by leofeyer** [comment]
Also, we should support the `+` sign in the local name.

**Comment by discordier** [comment]
There are way more valid characters. Maybe we should also allow them and pass the matched parts through filter email.

**Comment by leofeyer** [comment]
There are two approaches IMHO and we should choose one before going on with the discussion:

**1. Regex only**

If we only use a regex to match e-mail addresses in a string, it should be as exact as possible and reasonable. We should probably go for:

```
[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?
```

as explained here: http://www.regular-expressions.info/email.html

**2. Regex with additional validation**

Or we use the regex only to match substrings that might be an e-mail address and validate our findings afterwards. In this case the regex should be as simple as possible, probably just:

```
[^@]{1,64}@[^@]{1,255}
```

or even:

```
/.+@.+\..+/i
```

Then when rebuilding the string, these findings need to be validated by any e-mail validation class out there or even by the PHP function `filter_var()`.

**3. Pareto efficient solution**

This would be keeping our current solution with the enhanced length and support for the plus sign (+). It will work for the majority of use cases while keeping the complexity at a reasonable level.

---

Also see:
http://www.regular-expressions.info/email.html
http://www.linuxjournal.com/article/9585
http://davidcel.is/posts/stop-validating-email-addresses-with-regex/

**Comment by leofeyer** [comment]
If we should decide to go for number 2, here is a library we could use: https://github.com/egulias/EmailValidator

**Comment by leofeyer** [comment]
Fixed in fc2e101ec2c0e0bb18b48a85af33980309e93f92.

**Comment by ausi** [comment]
If we want to support unicode characters, we would also have to change the encoding loop to support multibyte characters.

With the changes in fc2e101ec2c0e0bb18b48a85af33980309e93f92 `StringUtil::encodeEmail('föo@exämple.com')` breaks now and outputs `fÃ¶o@exÃ¤mple.com`.

**Comment by leofeyer** [comment]
Fixed in a6aa8e756ad78f15fc93d27c4ee8289510c96f98.
